Add migrating to 8.0 guide (#133)
smartdevicelink/sdl_core_guides: docs/Migrating to Newer SDL Versions/Migrating SDL Core 7.1 to 8.0/index.md

> # Migrating SDL Core 7.1 to 8.0

Reviewer: Another section should be added for 0205, maybe something like this:

## HMI Updates

### Avoid Custom button subscription in case HMI incompatibility

SDL Core 8.0.0 no longer automatically subscribes to `CUSTOM_BUTTON`. If an HMI supports soft buttons, it must include an entry for `CUSTOM_BUTTON` in its' button capabilities in order for mobile to receive OnButtonPress and OnButtonEvent notifications.

## Environment Updates

### Ubuntu Versions

SDL Core 8.0.0 no longer supports Ubuntu 16. Supported versions of SDL Core are Ubuntu 18.04 and Ubuntu 20.04.

### SSL versions

SDL Core 8.0.0 dropped support for libssl1.0. Developers should install libssl-dev instead of libssl1.0-dev

## Updates to CMAKE Build Configuration

### Removed Flag `ENABLE_HMI_PTU_DECRYPTION`

`ENABLE_HMI_PTU_DECRYPTION` was removed from the build configuration. Behaviors defined by ON/OFF options are now both supported without the need for this build flag.

### Boost Logger

The default logger is still using LOG4CXX but the option is now available to use Boost for the logger. When porting SDL Core to different linux environments, the LOG4CXX logger was known to cause dependency issues. Boost is offered as an alternative logger in hopes to make porting SDL Core to different environments easier.

```
set(LOGGER_NAME "LOG4CXX" CACHE STRING "Logging library to use (BOOST, LOG4CXX)")
```

## Updates to Configuration File smartDeviceLink.ini

### DefaultTimeoutCompensation

> This parameter was added to the smartDeviceLink.ini configuration to help compensate for timing issues that could occur during `BasicCommunication.OnResetTimeout`. This value is added to the DefaultTimeout parameter when calculating the RPC request timeout.

Reviewer: I think this section could use some more information, and be a bit clearer, as it isn't necessarily related to resetting the timeout.

Background: `DefaultTimeoutCompensation` is a value Core adds to each timeout when waiting for a message response. This value is to compensate for the time a message spends "in the wire". For example, when Mobile requests an alert with a timeout of 5000ms, Core processes this message before the HMI, so if it set a timer for just 5000ms, it would expire and mobile would be sent GENERIC_ERROR before the HMI's TIMEOUT response is processed by Core. In prior releases, Core hardcoded adding additional time to some requests that may receive a TIMEOUT response. This new INI option `DefaultTimeoutCompensation` replaces the hardcoded timeout extensions with a configurable value that applies to all requests.

```suggestion
This parameter was added to the smartDeviceLink.ini configuration to compensate for transfer and processing time of requests. This value is added to the DefaultTimeout parameter when calculating the RPC request timeout. Previously, specific requests such as Alert were hardcoded to extend their default timeout, now timeout compensation is configurable and applied to all requests.
```

			<description>The ID of the application requesting this button usubscription. </description>
        </param>
		
        <param name="buttonName" type="ButtonName" mandatory="true">
            <description>Name of the button to subscribe.</description>
        </param>
</function>

<function name="SubscribeButton" messagetype="response"> </function>
```

## Buttons.UnsubscribeButton

`Buttons.UnsubscribeButton` request and response were added to allow SDL Core to request the HMI unsubscribes an application from a specific button.

```
<function name="UnsubscribeButton" messagetype="request">
        <description>
            Unsubscribes from buttons.            
        </description>
        
        <param name="appID" type="Integer" mandatory="true">
            <description>The ID of the application requesting this button unsubscription. </description>
        </param>
        
        <param name="buttonName" type="ButtonName" mandatory="true">
            <description>Name of the button to unsubscribe.</description>
        </param>
    </function>

<function name="UnsubscribeButton" messagetype="response"></function>

```

> 

Reviewer 2: Should probably include a section for the behavior changes for the phone call event here:

## OnEventChanged (PHONE_CALL)

The behavior of the `PHONE_CALL` event was changed to only affect the `audioStreamingState` of an app. Rather than automatically deactivating the active app, SDL Core will now only change the `audioStreamingState` of all apps to `NOT_AUDIBLE` when `BC.OnEventChanged(PHONE_CALL, active=true)` is sent, leaving each app's `hmiLevel` unchanged. This allows the HMI to start a phone call in the background without leaving the app screen, if desired.

The HMI can still control the `hmiLevel` of the app during a phone call event by sending `BC.OnAppDeactivated(appID)` and `BC.OnAppActivated(appID)` where appropriate.

# Migrating SDL Core 7.1 to 8.0

## Environment Updates

### Ubuntu Versions

SDL Core 8.0.0 no longer supports Ubuntu 16. Supported versions of SDL Core are Ubuntu 18.04 and Ubuntu 20.04.

> ### SSL versions

Reviewer 3: ```suggestion
### SSL Versions
```

> SDL Core 8.0.0 dropped support for libssl1.0. Developers should install libssl-dev instead of libssl1.0-dev

Reviewer 4: ```suggestion
SDL Core 8.0.0 dropped support for libssl1.0. Developers should install `libssl-dev` instead of `libssl1.0-dev`.
```

> SDL Core 8.0.0 dropped support for libssl1.0. Developers should install libssl-dev instead of libssl1.0-dev

Reviewer 3: ```suggestion
SDL Core 8.0.0 dropped support for libssl1.0. Developers should install libssl-dev instead of libssl1.0-dev.
```

## Updates to CMAKE Build Configuration

### Removed Flag `ENABLE_HMI_PTU_DECRYPTION`

`ENABLE_HMI_PTU_DECRYPTION` was removed from the build configuration. Behaviors defined by ON/OFF options are now both supported without the need for this build flag.

### Boost Logger

> The default logger is still using LOG4CXX but the option is now available to use Boost for the logger. When porting SDL Core to different linux environments, the LOG4CXX logger was known to cause dependency issues. Boost is offered as an alternative logger in hopes to make porting SDL Core to different environments easier.

Reviewer 3: ```suggestion
The default logger is still using LOG4CXX but the option is now available to use Boost for the logger. When porting SDL Core to different Linux environments, the LOG4CXX logger was known to cause dependency issues. Boost is offered as an alternative logger in hopes of making porting SDL Core to different environments easier.
```

Author: Updated to "hopes of making"

```
set(LOGGER_NAME "LOG4CXX" CACHE STRING "Logging library to use (BOOST, LOG4CXX)")
```

## Updates to Configuration File smartDeviceLink.ini

### DefaultTimeoutCompensation

> This parameter was added to the smartDeviceLink.ini configuration to compensate for transfer and processing time of requests. This value is added to the DefaultTimeout parameter when calculating the RPC request timeout. Previously, specific requests such as Alert were hardcoded to extend their default timeout, now timeout compensation is configurable and applied to all requests.

Reviewer 3: ```suggestion
This parameter was added to the smartDeviceLink.ini configuration to compensate for transfer and processing time of requests. This value is added to the `DefaultTimeout` parameter when calculating the RPC request timeout. Previously, specific requests such as Alert were hardcoded to extend their default timeout, now timeout compensation is configurable and applied to all requests.
```

```
; Extra time to compensate default timeout due to external delays
DefaultTimeoutCompensation = 1000
```

### Icons Storage Folder

The default parameter `AppIconsFolder` was updated to use a directory named “icons”. This value used to be “storage”.

```
; Specify a dedicated folder, as old files in this folder can be automatically removed
AppIconsFolder = icons
```

## HMI Behavior Changes

> ### Avoid Custom button subscription in case HMI incompatibility

Reviewer 3: ```suggestion
### Avoid Custom Button Subscription in Case HMI Incompatibility
```

> SDL Core 8.0.0 no longer automatically subscribes to `CUSTOM_BUTTON`. If an HMI supports soft buttons, it must include an entry for `CUSTOM_BUTTON` in its' button capabilities in order for mobile to receive OnButtonPress and OnButtonEvent notifications.

Reviewer 3: ```suggestion
SDL Core 8.0.0 no longer automatically subscribes to `CUSTOM_BUTTON`. If an HMI supports soft buttons, it must include an entry for `CUSTOM_BUTTON` in its button capabilities in order for mobile to receive `OnButtonPress` and `OnButtonEvent` notifications.
```

### OnEventChanged (PHONE_CALL)

The behavior of the `PHONE_CALL` event was changed to only affect the `audioStreamingState` of an app. Rather than automatically deactivating the active app, SDL Core will now only change the `audioStreamingState` of all apps to `NOT_AUDIBLE` when `BC.OnEventChanged(PHONE_CALL, active=true)` is sent, leaving each app's `hmiLevel` unchanged. This allows the HMI to start a phone call in the background without leaving the app screen, if desired.

The HMI can still control the `hmiLevel` of the app during a phone call event by sending `BC.OnAppDeactivated(appID)` and `BC.OnAppActivated(appID)` where appropriate.

## HMI API Updates

> ## Buttons.SubscribeButton

Reviewer 4: ```suggestion
### Buttons.SubscribeButton
```

> `Buttons.OnButtonSubscription` notification was replaced by `Buttons.SubscribeButton` Request and Response.

Reviewer 3: ```suggestion
`Buttons.OnButtonSubscription` notification was replaced by `Buttons.SubscribeButton` request and response.
```

```xml
<function name="SubscribeButton" messagetype="request">
        <description>
            Subscribes to buttons.            
        </description>
		
	    <param name="appID" type="Integer" mandatory="true">

> 			<description>The ID of the application requesting this button usubscription. </description>

Reviewer 3: ```suggestion
			<description>The ID of the application requesting this button subscription. </description>
```

        </param>
		
        <param name="buttonName" type="ButtonName" mandatory="true">
            <description>Name of the button to subscribe.</description>
        </param>
</function>

<function name="SubscribeButton" messagetype="response"> </function>
```

> ## Buttons.UnsubscribeButton

Reviewer 4: ```suggestion
### Buttons.UnsubscribeButton
```

> `Buttons.UnsubscribeButton` request and response were added to allow SDL Core to request the HMI unsubscribes an application from a specific button.

Reviewer 3: ```suggestion
`Buttons.UnsubscribeButton` request and response were added to allow SDL Core to request that the HMI unsubscribes an application from a specific button.
```

```
<function name="UnsubscribeButton" messagetype="request">
        <description>
            Unsubscribes from buttons.            
        </description>
        
        <param name="appID" type="Integer" mandatory="true">
            <description>The ID of the application requesting this button unsubscription. </description>
        </param>
        
        <param name="buttonName" type="ButtonName" mandatory="true">
            <description>Name of the button to unsubscribe.</description>
        </param>
    </function>

<function name="UnsubscribeButton" messagetype="response"></function>

```

> ## Restructuring OnResetTimeout 

Reviewer 4: ```suggestion
### Restructuring OnResetTimeout 
```

> `UI.OnResetTimeout` and `TTS.OnResetTimeout` were removed in place of using a broader rpc, `BasicCommunication.OnResetTimeout`.

Reviewer 3: ```suggestion
`UI.OnResetTimeout` and `TTS.OnResetTimeout` were removed in place of using a broader RPC, `BasicCommunication.OnResetTimeout`.
```

This updated `OnResetTimeout` RPC can be used across all interfaces for all request functions.

The parameters in the notification have also changed.

> The parameter `requestID` is used instead of appID to identify which specific request should have its timeout extended.

Reviewer 3: ```suggestion
The parameter `requestID` is used instead of `appID` to identify which specific request should have its timeout extended.
```

> The parameter `methodName` should include the interface name and the rpc. For example: `”TTS.Speak”`.

Reviewer 3: ```suggestion
The parameter `methodName` should include the interface name and the RPC. For example: `”TTS.Speak”`.
```

Reviewer 3: Additional question: should this be `TTS.Speak` or are the double quotes intentional?

> The parameter `resetPeriod` allows the HMI to specify how long Core should delay the application request’s timeout.

Reviewer 4: ```suggestion
The parameters in the notification have also changed:
- The parameter `requestID` is used instead of appID to identify which specific request should have its timeout extended.
- The parameter `methodName` should include the interface name and the rpc. For example: `”TTS.Speak”`.
- The parameter `resetPeriod` allows the HMI to specify how long Core should delay the application request’s timeout.
```
